src/lib/src/component/navigation/navigation.component.ts

```diff
@@ -46,17 +49,36 @@ export class NavigationComponent implements OnInit {
     this.subscribeToEvents();
   }
 
+  ngOnDestroy(): void {
+    this.stopPolling.next();
+    this.stopPolling.complete();
+  }
+
   retrieveWorkspaceRoot(): Promise<Workspace | undefined> {
     return this.persistenceService.listFiles().then(element => {
       this.setWorkspace(new Workspace(element));
       this.uiState.setExpanded(element.path, true);
+      this.updateTestStates();
       return this.workspace;
     }).catch(() => {
       this.errorMessage = 'Could not retrieve workspace!';
       return undefined;
     });
   }
 
+  private updateTestStates(): void {
+    this.stopPolling.next();
```

Comment:
I don't immediately get why the polling has to be stopped here. Is the subscription refreshed (only in setWorkspace)? Subscription, polling and stop polling seem to need some comments, so that (even) I understand their working ...

Reply:
Right, that's a bit convoluted, I guess. The gist of it is, when the workspace is refreshed, it gets thrown out completely. Unfortunately, the `Workspace` class is not well encapsulated, at all, so there can, and therefore _will_ be `WorkspaceElement` instances that remain referenced somewhere after a refresh. In particular, the polling tasks update the test state using a direct reference to the corresponding workspace element. Therefore, they should be stopped on a refresh, because they will a) cause traffic senselessly as the status updates won't be visible, and b) so the references to those workspace elements are removed and garbage collection can reclaim them (avoiding a memory leak).

… I actually started to refactor this before I implemented it the way it is now, trying to better encapsulate class `Workspace`, and refreshing its state rather than throwing it out, completely. That quickly proved to be much more involved than I had hoped, though, so I stopped following that road. I think it would still make sense to refactor this, as it will only get worse…